internal/gengapic/genrest.go

```diff
@@ -245,36 +311,39 @@ func (g *generator) generateQueryString(m *descriptor.MethodDescriptorProto) {
 	}
 	sort.Strings(fields)
 
-	p("queryParamStrs := []string{}")
-	// TODO(dovs): make this work for nested fields
-	for _, f := range fields {
-		field := queryParams[f]
-		f = snakeToCamel(f)
+	requestObject := "req"
+
+	p("params := url.Values{}")
+	for _, path := range fields {
+		field := queryParams[path]
+		accessor := buildAccessor(path, false)
 		if field.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
 			// It's a slice, so check for nil
-			p("if req.Get%s() != nil {", f)
+			p("if %s%s != nil {", requestObject, accessor)
+		} else if field.GetProto3Optional() {
+			// Split right before the raw access
+			elts := strings.Split(path, ".")
+			elts = elts[:len(elts)-1]
+			p("if %[1]s%[2]s != nil && %[1]s%[3]s != nil {", requestObject, buildAccessor(strings.Join(elts, "."), false), buildAccessor(path, true))
```

Comment:
Nit: consider wrapping `buildAccessor(str, false)` and `buildAccesor(str, true)` in methods that convey what the boolean toggles i.e.

```go
func getter(s string) string {
  return buildAccessor(s, true)
}

func directAccess(s string) string {
  return buildAccessor(s, false)
}
```

Honestly might be an over reaction, but might read a little better 🤷🏻‍♂️

Reply:
That's really helpful, actually. One thing I particularly miss from python/lisp is keyword params, which I think could have made calls to `buildAccessor` a little more clear.
Something like
```go
// Strawman syntax
func buildAccessor(s string, rawFinal bool = false) { ... }

buildAccessor(path)
buildAccessor(path, rawFinal=true)
```